src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -70,589 +91,522 @@ public async ValueTask DisposeAsync()
-    /// when sending on the wire
-    /// </param>
+    /// <param name="sizeHint">The minimum number of items to ensure space for in <see cref="_array"/>.</param>
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private void EnqueueCommand(bool success, int trim = 0)
+    private void CheckBufferAndEnsureCapacity(int sizeHint)
     {
-        var size = (int)_pipeWriter.UnflushedBytes;
-        if (size == 0)
+        var array = _array;
+
+        if (array is null)
         {
-            // no unflushed bytes means no command was produced
-            _flushTask = null;
-            return;
+            ThrowObjectDisposedException();
         }
 
-        if (success)
+        if (sizeHint < 0)
         {
-            Interlocked.Add(ref _counter.PendingMessages, 1);
+            ThrowArgumentOutOfRangeExceptionForNegativeSizeHint();
         }
 
-        _queuedCommandsWriter.TryWrite(new QueuedCommand(Size: size, Trim: success ? trim : size));
-        var flush = _pipeWriter.FlushAsync();
-        _flushTask = flush.IsCompletedSuccessfully ? null : flush.AsTask();
-    }
-}
-
-internal sealed class PriorityCommandWriter : IAsyncDisposable
-{
-    private readonly NatsPipeliningWriteProtocolProcessor _natsPipeliningWriteProtocolProcessor;
-    private int _disposed;
+        if (sizeHint == 0)
+        {
+            sizeHint = 1;
+        }
 
-    public PriorityCommandWriter(ISocketConnection socketConnection, NatsOpts opts, ConnectionStatsCounter counter, Action<PingCommand> enqueuePing)
-    {
-        CommandWriter = new CommandWriter(opts, counter, enqueuePing, overrideCommandTimeout: TimeSpan.MaxValue);
-        _natsPipeliningWriteProtocolProcessor = CommandWriter.CreateNatsPipeliningWriteProtocolProcessor(socketConnection);
+        if (sizeHint > array!.Length - _index)
+        {
+            ResizeBuffer(sizeHint);
+        }
     }
 
-    public CommandWriter CommandWriter { get; }
-
-    public async ValueTask DisposeAsync()
+    /// <summary>
+    /// Resizes <see cref="_array"/> to ensure it can fit the specified number of new items.
+    /// </summary>
+    /// <param name="sizeHint">The minimum number of items to ensure space for in <see cref="_array"/>.</param>
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private void ResizeBuffer(int sizeHint)
     {
-        if (Interlocked.Increment(ref _disposed) == 1)
+        var minimumSize = (uint)_index + (uint)sizeHint;
+
+        // The ArrayPool<T> class has a maximum threshold of 1024 * 1024 for the maximum length of
+        // pooled arrays, and once this is exceeded it will just allocate a new array every time
+        // of exactly the requested size. In that case, we manually round up the requested size to
+        // the nearest power of two, to ensure that repeated consecutive writes when the array in
+        // use is bigger than that threshold don't end up causing a resize every single time.
+        if (minimumSize > 1024 * 1024)
```

Comment:
<3 the exposition here! Great breakdown of the why.

Reply:
not me unfortunately 😅 originally taken from the community toolkit (as mentioned in NatsBufferWriter.cs class). I'll make sure to put an attribution comment later.